test/TCPConnectionTests.cc

```diff
@@ -0,0 +1,254 @@
+            //Send something
+            server_conn->send_message(reinterpret_cast<const uint8_t*>(server_to_client.c_str()),
+                                     server_to_client.length());
+        }
+
+        for (int i = 0; i < NUMBER_OF_MESSAGES; ++i) {
+            //Recieve something
+            BytesBuffer message_received = server_conn->recv_message();
+            std::string recv_message(reinterpret_cast<char*>(const_cast<uint8_t*>(message_received.data())));
+            ASSERT_EQ(0, recv_message.compare(client_to_server));
+        }
+    });
+
+    comm::ConnClient conn_client({"localhost", SERVER_PORT_MULTIPLE});
+
+    barrier.wait();
+
+    auto connection = conn_client.connect();
+
+    for (int i = 0; i < NUMBER_OF_MESSAGES; ++i) {
+        // Send something
+        connection->send_message(reinterpret_cast<const uint8_t*>(client_to_server.c_str()),
+                                 client_to_server.length());
+    }
+
+    for (int i = 0; i < NUMBER_OF_MESSAGES; ++i) {
+        // Receive something
+        BytesBuffer message_received = connection->recv_message();
+        std::string recv_message(reinterpret_cast<char*>(const_cast<uint8_t*>(message_received.data())));
+        ASSERT_EQ(0, recv_message.compare(server_to_client));
+    }
+
+    server_thread.join();
+}
+
+// Server accepts connection and then goes down, client tries to recv.
+TEST(TCPConnectionTests, ServerShutdownRecv)
+{
+    Barrier barrier(2);
+
+    std::thread server_thread([&]()
+    {
+        comm::ConnServer server(SERVER_PORT_INTERCHANGE);
+
+        barrier.wait();
+
+        server.accept();
+    });
+
+    comm::ConnClient conn_client({"localhost", SERVER_PORT_INTERCHANGE});
+
+    barrier.wait();
+
+    auto connection = conn_client.connect();
+
+    server_thread.join(); // Here the server will close the port.
+
+    ASSERT_THROW(
+        connection->recv_message(),
+        comm::Exception
+    );
+}
+
+TEST(TCPConnectionTests, SendArrayInts)
+{
+    int arr[10] = {22, 568, 254, 784, 452, 458, 235, 124, 1425, 1542};
+
+    Barrier barrier(2);
+
+    std::thread server_thread([&]()
+    {
+        comm::ConnServer server(SERVER_PORT_INTERCHANGE);
+
+        barrier.wait();
+
+        auto server_conn = server.accept();
+
+        server_conn->send_message(reinterpret_cast<const uint8_t*>(arr), sizeof(arr));
+    });
+
+    comm::ConnClient conn_client({"localhost", SERVER_PORT_INTERCHANGE});
+
+    barrier.wait();
+
+    auto connection = conn_client.connect();
+
+    BytesBuffer message_received = connection->recv_message();
+
+    const int* arr_recv = reinterpret_cast<const int*>(message_received.data());
+    for (int i = 0; i < 10; ++i) {
+        ASSERT_EQ(arr[i], arr_recv[i]);
+    }
+
+    server_thread.join();
+}
+
+TEST(TCPConnectionTests, MoveCopy)
+{
+    comm::TCPConnection a;
+    comm::TCPConnection server_conn;
+    server_conn = std::move(a); // Testing copy with move works
```

Comment:
this is a test to check the semantics compile and the most basic case works. Most of other tests rely on the move constructor working, so they will fail if something is off. This test is more of a debugging thing to signal if there is an issue with the most basic case (which may even fail when compiling the test).

Reply:
> may even fail when compiling the test

Yeah, if I'm reading this correctly the only way this can fail is if it fails to compile, which suggests to me that it may better suited to a compile time check (eg. `static_assert`) than a unit test.  But not a big deal either way.